pkg/datastore/memory.go

```diff
@@ -0,0 +1,130 @@
+package datastore
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"sync"
+	"time"
+
+	v1 "github.com/metal-stack/masterdata-api/api/v1"
+)
+
+type memoryDatastore[E Entity] struct {
+	lock     sync.RWMutex
+	entities map[string]E
+	entity   string
+	log      *slog.Logger
+}
+
+func NewMemory[E Entity](log *slog.Logger, e E) Storage[E] {
+	entity := e.JSONField()
+	return &memoryDatastore[E]{
+		lock:     sync.RWMutex{},
+		entities: make(map[string]E),
+		entity:   entity,
+		log:      log,
+	}
+}
+
+// Create implements Storage.
+func (m *memoryDatastore[E]) Create(ctx context.Context, ve E) error {
+	m.log.Debug("create", "entity", m.entity, "value", ve)
+	meta := ve.GetMeta()
+	if meta == nil {
+		return fmt.Errorf("create of type:%s failed, meta is nil", m.entity)
+	}
+
+	id := ve.GetMeta().Id
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	_, ok := m.entities[id]
+	if ok {
+		return NewDuplicateKeyError(fmt.Sprintf("an entity of type:%s with the id:%s already exists", m.entity, id))
+	}
+	m.entities[id] = ve
+	return nil
+}
+
+// Delete implements Storage.
+func (m *memoryDatastore[E]) Delete(ctx context.Context, id string) error {
+	m.log.Debug("delete", "entity", m.entity, "id", id)
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	_, ok := m.entities[id]
+	if !ok {
+		return NewNotFoundError(fmt.Sprintf("not found: delete of %s with id %s", m.entity, id))
+	}
+	delete(m.entities, id)
+
+	return nil
+}
+
+// DeleteAll implements Storage.
+func (m *memoryDatastore[E]) DeleteAll(ctx context.Context, ids ...string) error {
+	panic("unimplemented")
```

Comment:
Do we want to implement this? Should be fairly easy

Reply:
done